# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: parser/html/nsHtml5Tokenizer.h
# Commit: 9a8abd87cc79
# Full Hash: 9a8abd87cc7935f29b94248c1a6f8203faa14403
# Author: Henri Sivonen <hsivonen@hsivonen.fi>
# Date: 2021-12-08 21:43:06
# Regressor Bug: 1701828
# File Overlap Count: 1
# Description:
#   Bug 1701828 - meta charset rewrite. r=smaug
#   
#   Implements https://github.com/whatwg/html/issues/6962 . Improves performance
#   when <meta charset> occurs in head but after the first kilobyte and aligns
#   behavior better with WebKit and Blink.
# ==============================================================================

diff --git a/parser/html/nsHtml5Tokenizer.h b/parser/html/nsHtml5Tokenizer.h
--- a/parser/html/nsHtml5Tokenizer.h
+++ b/parser/html/nsHtml5Tokenizer.h
@@ -47,17 +47,16 @@
 #include "nsIContent.h"
 #include "nsTraceRefcnt.h"
 
 class nsHtml5StreamParser;
 
 class nsHtml5AttributeName;
 class nsHtml5ElementName;
 class nsHtml5TreeBuilder;
-class nsHtml5MetaScanner;
 class nsHtml5UTF16Buffer;
 class nsHtml5StateSnapshot;
 class nsHtml5Portability;
 
 class nsHtml5Tokenizer {
  private:
   static const int32_t DATA_AND_RCDATA_MASK = ~1;
 
@@ -265,16 +264,17 @@ class nsHtml5Tokenizer {
   int32_t candidate;
   int32_t charRefBufMark;
 
  protected:
   int32_t value;
 
  private:
   bool seenDigits;
+  bool suspendAfterCurrentNonTextToken;
 
  protected:
   int32_t cstart;
 
  private:
   nsHtml5String publicId;
   nsHtml5String systemId;
   autoJArray<char16_t, int32_t> strBuf;
@@ -447,16 +447,19 @@ class nsHtml5Tokenizer {
   void bogusDoctypeWithoutQuirks();
   void handleNcrValue(int32_t returnState);
 
  public:
   void eof();
 
  private:
   void emitDoctypeToken(int32_t pos);
+  void suspendIfRequestedAfterCurrentNonTextToken();
+  void suspendAfterCurrentTokenIfNotInText();
+  bool suspensionAfterCurrentNonTextTokenPending();
 
  protected:
   inline char16_t checkChar(char16_t* buf, int32_t pos) { return buf[pos]; }
 
  public:
   bool internalEncodingDeclaration(nsHtml5String internalCharset);
 
  private: